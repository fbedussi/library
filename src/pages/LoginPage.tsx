import { Field, Form, Formik } from 'formik'
import React from 'react'
import { useTranslation } from 'react-i18next'
import { useDispatch, useSelector } from 'react-redux'
import { Redirect } from 'react-router-dom'
import styled from 'styled-components'

import { PageWrapper } from '../components/CommonComponents'
import { pxToRem } from '../libs/styles'
import authActions from '../store/auth/actions'
import { selectUserId } from '../store/auth/selectors'
import { Button, FormControlLabel, Switch, TextField } from '../styleguide'
import theme from '../styleguide/theme'

const LoginPageWrapper = styled(PageWrapper)`
	justify-content: center;
`;

const LoginForm = styled(Form)`
	margin: 0 auto;
	max-width: 500px;
	display: flex;
	flex-direction: column;
	margin-bottom: ${pxToRem(theme.spacing(2))}rem;

	> * {
		margin-bottom: ${pxToRem(theme.spacing(1))}rem;
	}
`;

const ButtonWrapper = styled.div`
	display: flex;
	width: 100%;
`;

const LoginPage: React.FC = () => {
	const userId = useSelector(selectUserId);
	const { t } = useTranslation();
	const dispatch = useDispatch();

	return userId ? (
		<Redirect to="/search" />
	) : (
		<LoginPageWrapper>
			<Formik
				initialValues={{
					username: '',
					password: '',
					rememberMe: false,
				}}
				onSubmit={values => {
					dispatch(authActions.login(values));
				}}
			>
				{() => {
					return (
						<LoginForm>
							<Field
								as={TextField}
								name="username"
								id="username"
								label={t('app.username')}
								variant="outlined"
							/>
							<Field
								as={TextField}
								name="password"
								id="password"
								type="password"
								label={t('app.password')}
								variant="outlined"
							/>
							<FormControlLabel
								control={<Field as={Switch} name="rememberMe" />}
								label={t('app.rememberMe')}
							/>
							<ButtonWrapper>
								<Button type="submit" color="primary" variant="contained">
									{t('app.login')}
								</Button>
							</ButtonWrapper>
						</LoginForm>
					);
				}}
			</Formik>
		</LoginPageWrapper>
	);
};

export default LoginPage;
