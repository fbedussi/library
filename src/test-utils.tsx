import React from 'react'
import { Provider } from 'react-redux'
import { BrowserRouter } from 'react-router-dom'

import {
  ByRoleMatcher, ByRoleOptions, render as rtlRender,
  screen as rtlScreen
} from '@testing-library/react'

import { RootState } from './model/model'
import { getTestStore } from './store'

function render(
	ui: JSX.Element,
	{
		initialState = {},
		store = getTestStore(initialState),
		dispatch,
		route = '/',
		...renderOptions
	} = {} as {
		initialState?: Partial<RootState>;
		store?: any;
		dispatch?: any;
		route?: string;
	},
) {
	const Wrapper: React.FC<{ children?: JSX.Element }> = ({ children }) => {
		return (
			<Provider store={dispatch ? { ...store, dispatch } : store}>
				<BrowserRouter>{children}</BrowserRouter>
			</Provider>
		);
	};
	return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

const screen = {
	...rtlScreen,
	getByRole: <T extends HTMLElement>(
		text: ByRoleMatcher,
		options?: ByRoleOptions | undefined,
		waitForElementOptions?: unknown,
	) => rtlScreen.getByRole(text, options, waitForElementOptions) as T,
};

// re-export everything
export * from '@testing-library/react';
// override render method
export { render, screen };
