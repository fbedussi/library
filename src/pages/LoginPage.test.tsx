import React from 'react'
import { act } from 'react-dom/test-utils'

import { fireEvent, screen, waitFor } from '@testing-library/react'

import { render } from '../test-utils'
import LoginPage from './LoginPage'

jest.mock('../store/auth/actions', () => {
	return {
		login: (x: any) => ({ ...x, type: 'login' }),
	};
});

describe('LoginPage', () => {
	it('redirects to search if userId is populated', () => {
		const page = render(<LoginPage />, {
			initialState: { auth: { userId: 'foo' } },
		});
		expect(page).toMatchSnapshot();
	});

	it('renders', () => {
		const page = render(<LoginPage />);
		expect(page).toMatchSnapshot();
	});

	it('submits', () => {
		const dispatch = jest.fn();
		render(<LoginPage />, { dispatch });
		act(() => {
			fireEvent.change(screen.getByLabelText(/app.username/), {
				target: { value: 'foo' },
			});
		});
		act(() => {
			fireEvent.change(screen.getByLabelText(/app.password/), {
				target: { value: 'baz' },
			});
		});
		act(() => {
			fireEvent.click(screen.getByLabelText(/app.rememberMe/));
		});
		waitFor(() => {
			expect(dispatch).toBeCalledWith({
				username: 'foo',
				password: 'baz',
				rememberMe: true,
			});
		});
	});
});
