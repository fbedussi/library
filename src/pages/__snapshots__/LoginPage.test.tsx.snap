// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-bcXHqe sc-iBYQkv ljvCyv eqvGvc"
      >
        <form
          action="#"
          class="sc-ftTHYK gngptd"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="username"
              id="username-label"
            >
              app.username
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id="username"
                name="username"
                type="text"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-yjsfm1"
                >
                  <span>
                    app.username
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <div
            class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="password"
              id="password-label"
            >
              app.password
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id="password"
                name="password"
                type="password"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-yjsfm1"
                >
                  <span>
                    app.password
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <label
            class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
          >
            <span
              class="MuiSwitch-root MuiSwitch-sizeMedium css-julti5-MuiSwitch-root"
            >
              <span
                class="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary css-5ryogn-MuiButtonBase-root-MuiSwitch-switchBase"
              >
                <input
                  class="PrivateSwitchBase-input MuiSwitch-input css-1m9pwf3"
                  name="rememberMe"
                  type="checkbox"
                  value="false"
                />
                <span
                  class="MuiSwitch-thumb css-jsexje-MuiSwitch-thumb"
                />
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </span>
              <span
                class="MuiSwitch-track css-1yjjitx-MuiSwitch-track"
              />
            </span>
            <span
              class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
            >
              app.rememberMe
            </span>
          </label>
          <div
            class="sc-pyfCe hRSUzW"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="submit"
            >
              app.login
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-bcXHqe sc-iBYQkv ljvCyv eqvGvc"
    >
      <form
        action="#"
        class="sc-ftTHYK gngptd"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="username"
            id="username-label"
          >
            app.username
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="username"
              name="username"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  app.username
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="password"
            id="password-label"
          >
            app.password
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="password"
              name="password"
              type="password"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  app.password
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <label
          class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
        >
          <span
            class="MuiSwitch-root MuiSwitch-sizeMedium css-julti5-MuiSwitch-root"
          >
            <span
              class="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary css-5ryogn-MuiButtonBase-root-MuiSwitch-switchBase"
            >
              <input
                class="PrivateSwitchBase-input MuiSwitch-input css-1m9pwf3"
                name="rememberMe"
                type="checkbox"
                value="false"
              />
              <span
                class="MuiSwitch-thumb css-jsexje-MuiSwitch-thumb"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </span>
            <span
              class="MuiSwitch-track css-1yjjitx-MuiSwitch-track"
            />
          </span>
          <span
            class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
          >
            app.rememberMe
          </span>
        </label>
        <div
          class="sc-pyfCe hRSUzW"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="submit"
          >
            app.login
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
